lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+		public:    pub,
+	}, nil
+}
+
+func (s *kmsSigner) Public() crypto.PublicKey {
+	return s.public
+}
+
+func (s *kmsSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ctx := context.Background()
+
+	requestDigest := &kmspb.Digest{}
+	switch opts.HashFunc() {
+	case crypto.SHA256:
+		requestDigest.Digest = &kmspb.Digest_Sha256{
+			Sha256: digest,
+		}
+	case crypto.SHA512:
+		requestDigest.Digest = &kmspb.Digest_Sha512{
+			Sha512: digest,
+		}
+	default:
+		return nil, trace.BadParameter("unsupported hash func for GCP KMS signer: %v", opts.HashFunc())
+	}
+
+	resp, err := doGCPRequest(ctx, s.clock, s.kmsClient.AsymmetricSign, &kmspb.AsymmetricSignRequest{
+		Name:   s.keyID.keyVersionName,
+		Digest: requestDigest,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "error while attempting GCP KMS signing operation")
+	}
+	return resp.Signature, nil
+}
+
+type gcpKMSKeyID struct {
+	keyVersionName string
+}
+
+func (g gcpKMSKeyID) marshal() []byte {
+	var buf bytes.Buffer
+	buf.WriteString(gcpkmsPrefix)
+	buf.WriteString(g.keyVersionName)
+	return buf.Bytes()
+}
+
+func parseGCPKMSKeyID(key []byte) (gcpKMSKeyID, error) {
+	var keyID gcpKMSKeyID
+	if keyType(key) != types.PrivateKeyType_GCP_KMS {
+		return keyID, trace.BadParameter("unable to parse invalid GCP KMS key")
+	}
+	// strip gcpkms: prefix
+	keyID.keyVersionName = string(key[len(gcpkmsPrefix):])
+	return keyID, nil
+}
+
+// withTimeout is like context.WithTimeout, but with mockable time
+func withTimeout(ctx context.Context, clock clockwork.Clock, d time.Duration) (context.Context, func()) {
+	ctx, ctxCancel := context.WithCancel(ctx)
+	timer := clock.NewTimer(d)
+	cancel := func() {
+		ctxCancel()
+		timer.Stop()
+	}
+	go func() {
+		select {
+		case <-timer.Chan():
+		case <-ctx.Done():
+		}
+		cancel()
+	}()
+	return ctx, cancel
+}
+
+func retryWhilePending[reqType, optType, respType any](
+	ctx context.Context,
+	clock clockwork.Clock,
+	f func(context.Context, reqType, ...optType) (*respType, error),
+	req reqType,
+) (*respType, error) {
+	ctx, cancel := withTimeout(ctx, clock, defaultGCPPendingTimeout)
+	defer cancel()
+
+	ticker := clock.NewTicker(defaultGCPPendingRetryInterval)
+	defer ticker.Stop()
+	for {
+		resp, err := doGCPRequest(ctx, clock, f, req)
+		if err == nil {
+			return resp, nil
+		}
+		if !strings.Contains(err.Error(), "PENDING") {
+			return nil, trace.Wrap(err)
+		}
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-ticker.Chan():
+		}
+	}
+}
+
+func doGCPRequest[reqType, optType, respType any](
+	ctx context.Context,
+	clock clockwork.Clock,
+	f func(context.Context, reqType, ...optType) (*respType, error),
+	req reqType,
+) (*respType, error) {
+	ctx, cancel := withTimeout(ctx, clock, defaultGCPRequestTimeout)
+	defer cancel()
+	resp, err := f(ctx, req)
+	return resp, trace.Wrap(err)
+}
+
+// kmsClientI is a mockable interface for the GCP KMS client.
+type kmsClientI interface {
```

Comment:
```suggestion
type kmsClient interface {
```

In the vast majority of cases, we don't put an `I` suffix on our interface names. I'm thinking it's not a pattern we want to continue to encourage.

Reply:
I want the variable to be `kmsClient` so I went with `kmsService` for the interface type